// @uiw/react-textarea-code-editor/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
var _exportNames = {};
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _utils = require("./utils");

var _shortcuts = _interopRequireDefault(require("./shortcuts"));

var styles = _interopRequireWildcard(require("./styles"));

var _jsxRuntime = require("react/jsx-runtime");

var _SelectionText = require("./SelectionText");

Object.keys(_SelectionText).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _SelectionText[key]) return;
  exports[key] = _SelectionText[key];
});
var _excluded = ["prefixCls", "value", "padding", "minHeight", "placeholder", "language", "className", "style", "onChange"];

var _default = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _props$prefixCls = props.prefixCls,
      prefixCls = _props$prefixCls === void 0 ? 'w-tc-editor' : _props$prefixCls,
      _ = props.value,
      _props$padding = props.padding,
      padding = _props$padding === void 0 ? 10 : _props$padding,
      _props$minHeight = props.minHeight,
      minHeight = _props$minHeight === void 0 ? 16 : _props$minHeight,
      placeholder = props.placeholder,
      language = props.language,
      className = props.className,
      style = props.style,
      _onChange = props.onChange,
      other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  var _useState = (0, _react.useState)(props.value || ''),
      value = _useState[0],
      setValue = _useState[1];

  (0, _react.useEffect)(function () {
    return setValue(props.value || '');
  }, [props.value]);
  var textRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, function () {
    return textRef.current;
  });
  var contentStyle = {
    paddingTop: padding,
    paddingRight: padding,
    paddingBottom: padding,
    paddingLeft: padding
  };
  var htmlStr = (0, _react.useMemo)(function () {
    return (0, _utils.processHtml)("<pre aria-hidden=true><code " + (language ? "class=\"language-" + language + "\"" : '') + " >" + (0, _utils.htmlEncode)(String(value || '')) + "</code><br /></pre>");
  }, [value, language]);
  var preView = (0, _react.useMemo)(function () {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      style: (0, _extends2.default)({}, styles.editor, contentStyle, {
        minHeight: minHeight
      }),
      className: prefixCls + "-preview " + (language ? "language-" + language : ''),
      dangerouslySetInnerHTML: {
        __html: htmlStr
      }
    });
  }, // eslint-disable-next-line react-hooks/exhaustive-deps
  [prefixCls, language, htmlStr]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    style: (0, _extends2.default)({}, styles.container, style),
    className: prefixCls + " " + (className || ''),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("textarea", (0, _extends2.default)({
      autoComplete: "off",
      autoCorrect: "off",
      spellCheck: "false",
      autoCapitalize: "off"
    }, other, {
      placeholder: placeholder,
      onKeyDown: function onKeyDown(evn) {
        (0, _shortcuts.default)(evn);
        other.onKeyDown && other.onKeyDown(evn);
      },
      style: (0, _extends2.default)({}, styles.editor, styles.textarea, contentStyle, {
        minHeight: minHeight
      }, placeholder && !value ? {
        WebkitTextFillColor: 'inherit'
      } : {}),
      ref: textRef,
      onChange: function onChange(evn) {
        setValue(evn.target.value);
        _onChange && _onChange(evn);
      },
      className: prefixCls + "-text",
      value: value
    })), preView]
  });
});

exports.default = _default;
module.exports = exports.default;
